src/Psalm/Internal/Type/SimpleAssertionReconciler.php

```diff
@@ -2507,4 +2517,50 @@ function (Type\Union $t): bool {
             ? Type::getEmpty()
             : Type::getMixed();
     }
+
+    /**
+     * @param   0|1|2    $failed_reconciliation
+     */
+    private static function reconcileClassConstant(
+        Codebase $codebase,
+        string $class_constant_expression,
+        int &$failed_reconciliation
+    ) : Union {
+        if (strpos($class_constant_expression, '::') === false) {
+            $failed_reconciliation = 2;
+            return Type::getMixed();
+        }
+
+        [$class_name, $constant_pattern] = explode('::', $class_constant_expression, 2);
+
+        if (!$codebase->classlike_storage_provider->has($class_name)) {
+            $failed_reconciliation = 2;
+            return Type::getMixed();
+        }
+
+        $class_like_storage = $codebase->classlike_storage_provider->get($class_name);
+        $matched_class_constant_types = [];
+
+        foreach ($class_like_storage->constants as $constant => $class_constant_storage) {
+            if (!fnmatch($constant_pattern, $constant)) {
+                continue;
+            }
+
+            if (! $class_constant_storage->type || !$class_constant_storage->type->isSingle()) {
```

Comment:
Is there a reason to limit to Single types? You seem to add types in an array and to a big Union later so it should not be an issue to add every AtomicTypes in the array?

Reply:
Is it possible to have a constant of multiple types?
I've thought that a constant can only contain one type and thus I will enforce that here.

Reply:
something like [that](https://3v4l.org/AVdtA) could very well be inferred in Psalm with `float|int`

Reply:
I've changed the code accordingly